File: hosts/cpp/src/service/manifest.cpp
#include "service/manifest.hpp"

#include <nlohmann/json.hpp>

#include <fstream>
#include <sstream>
#include <stdexcept>

namespace qs::service
{

Manifest loadManifest(const std::string& path)
{
    std::ifstream in(path, std::ios::binary);
    if (!in)
        throw std::runtime_error("manifest not found: " + path);

    nlohmann::json j;
    try
    {
        in >> j;
    }
    catch (const std::exception& e)
    {
        throw std::runtime_error("manifest is not valid JSON (" + path + "): " + e.what());
    }

    if (!j.contains("module_path") || !j["module_path"].is_string())
        throw std::runtime_error("manifest is missing a string \"module_path\": " + path);
    if (!j.contains("log_level") || !j["log_level"].is_number_integer())
        throw std::runtime_error("manifest is missing an integer \"log_level\": " + path);

    Manifest m;
    m.module_path = j["module_path"].get<std::string>();
    m.log_level = j["log_level"].get<int>();
    m.commit = j.value("commit", std::string{});
    m.sdk_version = j.value("sdk_version", std::string{});
    m.mode = j.value("mode", std::string{});
    m.build_dir = j.value("build_dir", std::string{});
    return m;
}

}  // namespace qs::service

File: hosts/rust/src/service/manifest.rs
// Quackoscope host (Rust) -- service layer.
//
// manifest.json is where the module path, the log level, the SDK version and
// the SDK commit come from. None of the four is hardcoded anywhere in this
// source tree, and the handshake's sdk.version / sdk.commit are copied from
// here rather than asked of the running SDK.

use std::path::Path;

use serde_json::Value as Json;

#[derive(Debug, Clone, Default)]
pub struct Manifest {
    pub module_path: String,
    pub log_level: i64,
    pub sdk_version: String,
    pub commit: String,
    pub mode: String,
    pub build_dir: String,
}

pub fn load_manifest(path: &Path) -> Result<Manifest, String> {
    let text = std::fs::read_to_string(path)
        .map_err(|e| format!("manifest could not be read: {} ({e})", path.display()))?;

    let parsed: Json = serde_json::from_str(&text)
        .map_err(|e| format!("manifest is not valid JSON ({}): {e}", path.display()))?;

    let module_path = parsed
        .get("module_path")
        .and_then(Json::as_str)
        .ok_or_else(|| {
            format!(
                "manifest is missing a string \"module_path\": {}",
                path.display()
            )
        })?
        .to_string();

    let log_level = parsed.get("log_level").and_then(Json::as_i64).ok_or_else(|| {
        format!(
            "manifest is missing an integer \"log_level\": {}",
            path.display()
        )
    })?;

    let string_or_empty = |key: &str| {
        parsed
            .get(key)
            .and_then(Json::as_str)
            .unwrap_or_default()
            .to_string()
    };

    Ok(Manifest {
        module_path,
        log_level,
        sdk_version: string_or_empty("sdk_version"),
        commit: string_or_empty("commit"),
        mode: string_or_empty("mode"),
        build_dir: string_or_empty("build_dir"),
    })
}
